ED/AD: reactivating rotor furling and tailfin aerodynamics

ebranlard opened the pull request:
This pull request is not ready to be merged

**Feature or improvement description**
Reactivate the rotor and tail furling option in OpenFAST.
AeroDyn: 
- add a Tail Fin aero input file
- introduce a mesh for load/motions of the tail fin reference point
- compute aerodynamic loads at the tailfin reference point

ElastoDyn:
- change the ElastoDyn furling input file
- introduce a mesh for load/motions of the tail fin Center of mass
- compute kinematics at the tailfin center of mass 

GlueCode:
- Map loads/motions between AD and ED

**Impacted areas of the software**
AeroDyn/ElastoDyn/glue-code


**Checklist**
- [x] Delete point K from ElastoDyn
- [x] Optional: cleanup AeroDyn / ElastoDyn using newly added function CreatePointMesh
- [x] Documentation of  [AeroDyn changes](https://ebranlard-openfast.readthedocs.io/en/f-tailfin/source/user/aerodyn/input.html#tail-fin-aerodynamics),   [polar-based model](https://ebranlard-openfast.readthedocs.io/en/f-tailfin/source/user/aerodyn/theory_tailfin.html) and [ElastoDyn basics](https://ebranlard-openfast.readthedocs.io/en/f-tailfin/source/user/elastodyn/theory.html)

**Work for a future pull request**
- Implement unsteady slender body theory
- Implement induced velocity
- Implement tower influence
- Optional: VTK visualization
- Optional: additional outputs
- Documentation of USB model [here](https://ebranlard-openfast.readthedocs.io/en/f-tailfin/source/user/aerodyn/theory_tailfin.html)
- Test case

jjonkman commented on docs/source/user/aerodyn/input.rst:
```diff
@@ -754,3 +807,90 @@ See :numref:`ad_blade_geom`. Twist is shown in :numref:`ad_blade_local_cs` of :n
+    Comment
+    ======  General inputs =============================================================
+    1         TFinMod     - Tail fin aerodynamics model {0: none, 1: polar-based, 2: USB-based} (switch)
+    1.0       TFinArea    - Tail fin planform area (m^2) [used only when TFinMod=1]
```
It looks like the model also needs TFinChord for the tail fin pitching moment calculation.  Presumably this must be another input in the tail fin aerodynamics input file.

ebranlard replied:
That's correct, it was in the code but omitted in the documentation.

jjonkman commented on modules/nwtc-library/src/NWTC_Num.f90:
```diff
@@ -2246,6 +2250,49 @@ FUNCTION EulerExtractR16(M) result(theta)
+      ! this function creates a rotation matrix, M, from a 3-2-1 rotation
+      ! sequence of the 3 Euler angles, theta_z, theta_y, and theta_x, in radians.
+      ! M represents a change of basis (from global to local coordinates; 
+      ! not a physical rotation of the body). it is the inverse of EulerExtract
```
This would be the inverse of EulerExtractZYX(), if such a routine is written.  EulerExtract uses an XYZ rotation sequence.

jjonkman reviewed (commented):
Thanks, @ebranlard.  I'm excited to see tail fin aerodynamics and furling DOFs from FAST v7 re-introduced into OpenFAST!  Please find my initial review comments.  Overall I agree with the approach taken.

ebranlard commented on docs/source/user/elastodyn/coordsys.rst:
```diff
@@ -0,0 +1,126 @@
+The tail-furl bearing can be an ideal bearing with
+no friction by setting ``TFrlMod`` to 0; by setting
+``TFrlMod`` to 1, it also has a standard model that
+includes a linear spring, linear damper and Coulomb
```
The Coulomb damper is now removed from inputs and implementation.

ebranlard commented on docs/source/user/elastodyn/coordsys.rst:
```diff
@@ -0,0 +1,126 @@
+resisting motion equally in both directions once past
+the stop angle.
+
+A user-defined tail-furl spring and damper model
```
The code seems to support it, so I've left it. I've updated the documentation where necessary.

ebranlard commented on docs/source/user/elastodyn/coordsys.rst:
```diff
@@ -0,0 +1,126 @@
+           
+   Layout of a three-bladed, upwind, furling turbine: tail-furl structure.  
+   NOTE: The tail fin "CP" (center of pressure) parameters are now replaced by the location of the reference point.
+
```
I've added the description of the Rotor-furl coordinate system as given in the FAST 7 manual

ebranlard commented on docs/source/user/elastodyn/theory.rst:
```diff
@@ -0,0 +1,125 @@
+:math:`d_2 =z_n` and 
+:math:`d_3 =-y_n`.
+
+The following (partial) list of coordinate systems are defined by ElastoDyn:
```
I've now placed this in a "developer notes" section.

ebranlard commented:
Thank you for your review, I've addressed your comments, choosing to remove the Coulomb damping, but keeping the user-defined subroutines. 

I'll merge this with latest dev and create a different pull request in the future to address the following:
 - Implement unsteady slender body theory
 - Implement induced velocity
- Implement tower influence
- Optional: VTK visualization
- Optional: additional outputs
- Documentation of USB model 
- Test case
